{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metatranscriptome rRNA depletion and Library Prep Test\n",
    "### Yike Shen, Tessa R. Bloomquist, Abigail Rauso, Peter Meyn, Andrea A. Baccarelli\n",
    "### 12/06/2021\n",
    "\n",
    "YS, TRB, AR, AAB - Department of Environmental Health Sciences, Columbia University\\\n",
    "PM - Genome Technology Center, New York University\\\n",
    "Corresponding to: Yike Shen, yike.shen@columbia.edu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Due to ongoing challenging in rRNA depletion kits in the market in human stool shotgun metatranscriptome sequencing. We tested the following combinations of rRNA depletion and library prepration:\n",
    "- Illumina rRNA Depletion + Illumina Library Prepration\n",
    "- Illumina rRNA Depletion + NEB Library Prepration\n",
    "- NEB rRNA Depletion + Illumina Library Prepration\n",
    "- NEB rRNA Depletion + NEB Library Prepration\n",
    "\n",
    "Kits link:\n",
    "- Illumina Ribo-Zero Plus rRNA Depletion Kit: https://www.illumina.com/products/by-type/accessory-products/ribo-zero-plus-rrna-depletion.html\n",
    "- NEBNext® rRNA Depletion Kit (Bacteria): https://www.neb.com/products/e7850-nebnext-rrna-depletion-kit-bacteria#Product%20Information\n",
    "- Illumina Library Prepration (TruSeq Stranded mRNA): https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/truseq-stranded-mrna.html\n",
    "- NEBNext® Ultra™ II Directional RNA Library Prep Kit for Illumina®: https://www.neb.com/products/e7760-nebnext-ultra-ii-directional-rna-library-prep-kit-for-illumina#Product%20Information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.0 Sequencing depth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "896M\tIllDep-IllPrep-A_R1.fastq.gz\\\n",
    "943M\tIllDep-IllPrep-A_R2.fastq.gz\\\n",
    "957M\tIllDep-IllPrep-B_R1.fastq.gz\\\n",
    "987M\tIllDep-IllPrep-B_R2.fastq.gz\\\n",
    "663M\tIllDep-IllPrep-C_R1.fastq.gz\\\n",
    "715M\tIllDep-IllPrep-C_R2.fastq.gz\\\n",
    "1.9G\tIllDep-NEBPrep-A_R1.fastq.gz\\\n",
    "2.0G\tIllDep-NEBPrep-A_R2.fastq.gz\\\n",
    "1.9G\tIllDep-NEBPrep-B_R1.fastq.gz\\\n",
    "2.0G\tIllDep-NEBPrep-B_R2.fastq.gz\\\n",
    "1.2G\tIllDep-NEBPrep-C_R1.fastq.gz\\\n",
    "1.3G\tIllDep-NEBPrep-C_R2.fastq.gz\\\n",
    "680M\tNEBDep-IllPrep-A_R1.fastq.gz\\\n",
    "705M\tNEBDep-IllPrep-A_R2.fastq.gz\\\n",
    "712M\tNEBDep-IllPrep-B_R1.fastq.gz\\\n",
    "739M\tNEBDep-IllPrep-B_R2.fastq.gz\\\n",
    "876M\tNEBDep-IllPrep-C_R1.fastq.gz\\\n",
    "911M\tNEBDep-IllPrep-C_R2.fastq.gz\\\n",
    "1.9G\tNEBDep-NEBPrep-A_R1.fastq.gz\\\n",
    "2.0G\tNEBDep-NEBPrep-A_R2.fastq.gz\\\n",
    "1.7G\tNEBDep-NEBPrep-B_R1.fastq.gz\\\n",
    "1.8G\tNEBDep-NEBPrep-B_R2.fastq.gz\\\n",
    "1.8G\tNEBDep-NEBPrep-C_R1.fastq.gz\\\n",
    "1.9G\tNEBDep-NEBPrep-C_R2.fastq.gz\\\n",
    "\n",
    "Note: We allocated small reads to the test, the reads will be much deeper in a larger flowcell"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.0 Kneaddata to do metatranscriptome QC\n",
    "pseudocode example below, follow bioBakery Kneaddata tutorial to install dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --time=30:00:00\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --job-name=$jobname\n",
    "\n",
    "export OMP_NUM_THREADS=4\n",
    "conda init bash\n",
    "export CONDA3PATH=/$path/anaconda3\n",
    "module load Conda/3\n",
    "conda activate biobakery3\n",
    "\n",
    "kneaddata --input /$path/$Sample_R1.fastq.gz \\\n",
    "--input /$path/$Sample_R2.fastq.gz \\\n",
    "--reference-db /$path/bowtie2index \\\n",
    "--reference-db /$path/bowtie2transcriptome \\\n",
    "--reference-db /$path/silvadb \\\n",
    "--output /$path/kneaddataOutputPairedEnd_$sample \\\n",
    "--trimmomatic /opt/software/Trimmomatic/0.39-Java-1.8/ \\\n",
    "--trimmomatic-options=\"ILLUMINACLIP:/opt/software/Trimmomatic/0.39-Java-1.8/adapters/NexteraPE-PE.fa:2:30:10\"\n",
    "\n",
    "conda deactivate\n",
    "\n",
    "scontrol show job $SLURM_JOB_ID\n",
    "\n",
    "#Note, one kneaddata step is very long, I would recommend request at least 30 hours job run time\n",
    "#Note, preferable to run in scartch space, the mid files are very large, at least 20 times than the initial zipped size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.0 Run fastqc & multiqc "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir $directory\n",
    "export PATH=$PATH:/$path/FastQC/\n",
    "ls *fastq|while read line; do fastqc -o $directory/ -t 28 \"$line\";done\n",
    "\n",
    "multiqc ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.0 Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Raw reads (million)</th>\n",
       "      <th>minus adapter (million)</th>\n",
       "      <th>rRNA contamination (million)</th>\n",
       "      <th>Final reads (million)</th>\n",
       "      <th>Effective reads w duplication (%)</th>\n",
       "      <th>rRNA (%) to total</th>\n",
       "      <th>rRNA (%) to adapter removed</th>\n",
       "      <th>duplication rate (%)</th>\n",
       "      <th>Final non duplicated reads(million)</th>\n",
       "      <th>Final non duplicated %</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IllDepIllPrep_A</td>\n",
       "      <td>18.7</td>\n",
       "      <td>18.0</td>\n",
       "      <td>13.7</td>\n",
       "      <td>3.5</td>\n",
       "      <td>19.444444</td>\n",
       "      <td>73.262032</td>\n",
       "      <td>76.111111</td>\n",
       "      <td>67.4</td>\n",
       "      <td>1.1410</td>\n",
       "      <td>6.338889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IllDepNEBPrep_A</td>\n",
       "      <td>40.4</td>\n",
       "      <td>38.6</td>\n",
       "      <td>28.0</td>\n",
       "      <td>9.4</td>\n",
       "      <td>24.352332</td>\n",
       "      <td>69.306931</td>\n",
       "      <td>72.538860</td>\n",
       "      <td>67.4</td>\n",
       "      <td>3.0644</td>\n",
       "      <td>7.938860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NEBDepIllPrep_A</td>\n",
       "      <td>14.3</td>\n",
       "      <td>13.8</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.4</td>\n",
       "      <td>17.391304</td>\n",
       "      <td>76.923077</td>\n",
       "      <td>79.710145</td>\n",
       "      <td>76.0</td>\n",
       "      <td>0.5760</td>\n",
       "      <td>4.173913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEBDepNEBPrep_A</td>\n",
       "      <td>40.9</td>\n",
       "      <td>39.5</td>\n",
       "      <td>33.4</td>\n",
       "      <td>5.4</td>\n",
       "      <td>13.670886</td>\n",
       "      <td>81.662592</td>\n",
       "      <td>84.556962</td>\n",
       "      <td>60.0</td>\n",
       "      <td>2.1600</td>\n",
       "      <td>5.468354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IllDepIllPrep_B</td>\n",
       "      <td>17.4</td>\n",
       "      <td>16.8</td>\n",
       "      <td>7.7</td>\n",
       "      <td>8.6</td>\n",
       "      <td>51.190476</td>\n",
       "      <td>44.252874</td>\n",
       "      <td>45.833333</td>\n",
       "      <td>43.0</td>\n",
       "      <td>4.9020</td>\n",
       "      <td>29.178571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>IllDepNEBPrep_B</td>\n",
       "      <td>34.2</td>\n",
       "      <td>33.1</td>\n",
       "      <td>12.8</td>\n",
       "      <td>19.6</td>\n",
       "      <td>59.214502</td>\n",
       "      <td>37.426901</td>\n",
       "      <td>38.670695</td>\n",
       "      <td>47.0</td>\n",
       "      <td>10.3880</td>\n",
       "      <td>31.383686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NEBDepIllPrep_B</td>\n",
       "      <td>13.4</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.9</td>\n",
       "      <td>4.8</td>\n",
       "      <td>36.923077</td>\n",
       "      <td>58.955224</td>\n",
       "      <td>60.769231</td>\n",
       "      <td>46.0</td>\n",
       "      <td>2.5920</td>\n",
       "      <td>19.938462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NEBDepNEBPrep_B</td>\n",
       "      <td>31.8</td>\n",
       "      <td>30.8</td>\n",
       "      <td>17.0</td>\n",
       "      <td>13.3</td>\n",
       "      <td>43.181818</td>\n",
       "      <td>53.459119</td>\n",
       "      <td>55.194805</td>\n",
       "      <td>40.0</td>\n",
       "      <td>7.9800</td>\n",
       "      <td>25.909091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>IllDepIllPrep_C</td>\n",
       "      <td>16.0</td>\n",
       "      <td>15.5</td>\n",
       "      <td>12.8</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.903226</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>82.580645</td>\n",
       "      <td>66.0</td>\n",
       "      <td>0.6800</td>\n",
       "      <td>4.387097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>IllDepNEBPrep_C</td>\n",
       "      <td>32.2</td>\n",
       "      <td>31.2</td>\n",
       "      <td>25.8</td>\n",
       "      <td>4.5</td>\n",
       "      <td>14.423077</td>\n",
       "      <td>80.124224</td>\n",
       "      <td>82.692308</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1.4850</td>\n",
       "      <td>4.759615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NEBDepIllPrep_C</td>\n",
       "      <td>18.4</td>\n",
       "      <td>17.7</td>\n",
       "      <td>14.0</td>\n",
       "      <td>2.9</td>\n",
       "      <td>16.384181</td>\n",
       "      <td>76.086957</td>\n",
       "      <td>79.096045</td>\n",
       "      <td>81.0</td>\n",
       "      <td>0.5510</td>\n",
       "      <td>3.112994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NEBDepNEBPrep_C</td>\n",
       "      <td>41.9</td>\n",
       "      <td>40.7</td>\n",
       "      <td>35.5</td>\n",
       "      <td>4.4</td>\n",
       "      <td>10.810811</td>\n",
       "      <td>84.725537</td>\n",
       "      <td>87.223587</td>\n",
       "      <td>63.0</td>\n",
       "      <td>1.6280</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample  Raw reads (million)  minus adapter (million)  \\\n",
       "0   IllDepIllPrep_A                 18.7                     18.0   \n",
       "1   IllDepNEBPrep_A                 40.4                     38.6   \n",
       "2   NEBDepIllPrep_A                 14.3                     13.8   \n",
       "3   NEBDepNEBPrep_A                 40.9                     39.5   \n",
       "4   IllDepIllPrep_B                 17.4                     16.8   \n",
       "5   IllDepNEBPrep_B                 34.2                     33.1   \n",
       "6   NEBDepIllPrep_B                 13.4                     13.0   \n",
       "7   NEBDepNEBPrep_B                 31.8                     30.8   \n",
       "8   IllDepIllPrep_C                 16.0                     15.5   \n",
       "9   IllDepNEBPrep_C                 32.2                     31.2   \n",
       "10  NEBDepIllPrep_C                 18.4                     17.7   \n",
       "11  NEBDepNEBPrep_C                 41.9                     40.7   \n",
       "\n",
       "    rRNA contamination (million)  Final reads (million)  \\\n",
       "0                           13.7                    3.5   \n",
       "1                           28.0                    9.4   \n",
       "2                           11.0                    2.4   \n",
       "3                           33.4                    5.4   \n",
       "4                            7.7                    8.6   \n",
       "5                           12.8                   19.6   \n",
       "6                            7.9                    4.8   \n",
       "7                           17.0                   13.3   \n",
       "8                           12.8                    2.0   \n",
       "9                           25.8                    4.5   \n",
       "10                          14.0                    2.9   \n",
       "11                          35.5                    4.4   \n",
       "\n",
       "    Effective reads w duplication (%)  rRNA (%) to total  \\\n",
       "0                           19.444444          73.262032   \n",
       "1                           24.352332          69.306931   \n",
       "2                           17.391304          76.923077   \n",
       "3                           13.670886          81.662592   \n",
       "4                           51.190476          44.252874   \n",
       "5                           59.214502          37.426901   \n",
       "6                           36.923077          58.955224   \n",
       "7                           43.181818          53.459119   \n",
       "8                           12.903226          80.000000   \n",
       "9                           14.423077          80.124224   \n",
       "10                          16.384181          76.086957   \n",
       "11                          10.810811          84.725537   \n",
       "\n",
       "    rRNA (%) to adapter removed  duplication rate (%)  \\\n",
       "0                     76.111111                  67.4   \n",
       "1                     72.538860                  67.4   \n",
       "2                     79.710145                  76.0   \n",
       "3                     84.556962                  60.0   \n",
       "4                     45.833333                  43.0   \n",
       "5                     38.670695                  47.0   \n",
       "6                     60.769231                  46.0   \n",
       "7                     55.194805                  40.0   \n",
       "8                     82.580645                  66.0   \n",
       "9                     82.692308                  67.0   \n",
       "10                    79.096045                  81.0   \n",
       "11                    87.223587                  63.0   \n",
       "\n",
       "    Final non duplicated reads(million)  Final non duplicated %  \n",
       "0                                1.1410                6.338889  \n",
       "1                                3.0644                7.938860  \n",
       "2                                0.5760                4.173913  \n",
       "3                                2.1600                5.468354  \n",
       "4                                4.9020               29.178571  \n",
       "5                               10.3880               31.383686  \n",
       "6                                2.5920               19.938462  \n",
       "7                                7.9800               25.909091  \n",
       "8                                0.6800                4.387097  \n",
       "9                                1.4850                4.759615  \n",
       "10                               0.5510                3.112994  \n",
       "11                               1.6280                4.000000  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "pd.read_excel('metatranscriptome_test_results_Shen.xlsx', sheet_name='2021',header=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.0 Conclusions\n",
    "- NEB kit and illumina kit performs similar in rRNA depletion. Both not depleting well. \n",
    "- NEB library prep sequences much deeper than illumina library prep. If the allocated sequence deeper, illumina library prep was also able to sequence deep. \n",
    "- % of effective reads is sample dependent (B performs best)\n",
    "- rRNA depletion is still challenging, one option we decided is to do very deep sequencing and bioinformatically remove rRNA and other contamination (host and adapters)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
